Add LASSO without structure to LinearRegression notebook (#203)
equinor/iterative_ensemble_smoother: docs/source/kalman_objective.md

# Kalman-type ensemble approaches

How to think about Kalman-type ensemble based data assimilation:

- What does Kalman-type ensemble methods do?
- What is their ideal objective?
- How does the objective behave under statistical estimation?
- How to think about surrogate objectives?
- What is the modelling setup and what information is known a-priori?
- List common methods and discuss them in the context of the above.

What is _not_ provided, is an evaluation-comparison between methods.
Only how we _think_ they will behave, given what we know about statistics and our objective.
From the contents, it should however be clear how a simulation study could be constructed to compare methods.

> ## Gaussian conditional transport

Reviewer: The term `Gaussian conditional transport` is new to me, and a search provides no direct hits.
Is this a known term in statistics?

Author: It is a shortening of something like "Transporting a joint-Gaussian sample to a (Gaussian-)conditional sample". Perhaps better to spell it out if it is confusing.

Let $x$ and $y$ be random variables jointly multivariate Gaussian with mean and covariance

$$
\begin{bmatrix}
x \\
y
\end{bmatrix}
\sim \mathcal{N}\left(\begin{bmatrix}
\mu_x \\
\mu_y
\end{bmatrix}, \begin{bmatrix}
\Sigma_{x} & \Sigma_{xy} \\
\Sigma_{yx} & \Sigma_{y}
\end{bmatrix}\right)
$$

Define the "Kalman gain" as

$$
K = \Sigma_{xy}\Sigma_{y}^{-1}.
$$

Then, a multivariate sample $(x_i,y_i)$ is _transported_ to a sample from the conditional
$p(x|y)$, having observed $y$ as such, via the formula 

$$
x_i + K(y-y_i) \sim p(x|y).
$$

Note that $x|y$ retains Gaussianity as

$$
x | y \sim \mathcal{N}(\mu_x + K(y - \mu_y), \Sigma_{x} - K \Sigma_{yx}).
$$

For any distribution having a bijection between (the informative parts of) 
$x$ and $y$, there exist a similar mapping, transporting a sample $(x,y)$ to $x|y$.

> The EnKF and ES variants follows from writing $y=d=Hx+\epsilon$ where 

Reviewer: ```suggestion
The EnKF and ES variants follow from writing $y=d=Hx+\epsilon$ where 
```

$\epsilon\sim \mathcal{N}(0,\Sigma_{\epsilon})$, thus $\Sigma_y=H\Sigma_xH^T+\Sigma_{\epsilon}$.
Note that $y$ is common to use as $y=Hx$ but this is not the case in the preliminaries here.

> A point of confusion is that it is indeed _not_ the observation $d$ that is "perturbed" with noise,

Reviewer: Having mostly read Evensen, I am used to seeing $y = g(x)$ followed by $d=y + \epsilon$ where $f(\epsilon) = \mathcal{N}(0, \Sigma_d)$.
Do I understand correctly that this aligns well with how you think about "perturbed observations"?

As for the term itself, it kinda makes sense to me from a practical point of view.
Users of ert specify an observation $d$ by providing a value and an error $\epsilon$.
To create $D\in\mathcal{1\times N}$, we sample $N$ times from $\mathcal{0, \epsilon}$ and add that to $d$.
In this sense, we "perturbe" $d$ to create $D$.

What would be a better name for $D\in\mathcal{1\times N}$?

Author: From a practical point of view, it does not matter. From a theoretical point of view of understanding why these methods work, it matters. (Showcasing how I think about these methods was the goal of the markdown). We need a joint sample from $(x,d)$, and then want to "map" or "transport" this, given an observation say $d^\ast$, to $x|d^\ast$.

To achieve this, a joint sample $(x_i,d_i)$ is needed first. Here $d=h(x)+\epsilon$ or $d=g(x)+\epsilon$ depending on which notation you use. The deterministic model produces $h(x_i)$, but to achieve a correct sample from $d$ we need $d_i=h(x_i)+\epsilon_i$ where $\epsilon_i$ are samples from the observation uncertainty. The innovation is then $d^\ast - d_i$. But, since this is additive, and the noise is additive in $d$, and $\epsilon$ has a symmetric distribution about 0, we may perturb $d^\ast$ instead and compute $d_i^\ast=d^\ast+\epsilon_i$ and see that $d^\ast - d_i$ has the same distribution as $d_i^\ast - g(x_i)$.

The latter term is used extensively, where $d_i^\ast$ are the perturbed observations. In practice this works fine and are distributionally equivalent, but theoretically it adds another layer of confusion to what is really occurring.

---

Note that I do not intend to change language of methods, because I did not intend this as some reference theory document (this is why there are no references, among other things). My goal was to showcase how I personally think. Distribute statistical tools and understanding of methodology given that I think correctly, or become aware of faults in my own logic.

The very particular goal is to create an understanding for LASSO++ to get future PRs with new methodology reviewed and merged. Do we have suggestions/an idea of requirements for such PRs?

but rather, $d$ has variance $H\Sigma_xH^T+\Sigma_{\epsilon}$ and an $\epsilon$ is then added to the response $Hx$ to correctly sample from this distribution of $d$.
Because both $(d-d_i)$ and $d_i=Hx_i+\epsilon_i$ are additive, and $\epsilon$ is symmetric about 0, it does not matter if we add the noise to $d$ or to $Hx$ in the expression.
If any of these points were different, it would be evident that we are not perturbing observations.
Thus, it is incorrect and slightly confusing to say that we perturb the observations.
It is correct to say that we sample from the distribution of $d$, accounting from both prior uncertainty _and_ observation uncertainty.

> Note that the EnKF and ES is intended to work (and indeed do) when $d=h(x)+\epsilon$

Reviewer: ```suggestion
Note that the EnKF and ES are intended to work (and indeed do) when $d=h(x)+\epsilon$
```

The optimization can be done disjoint if marginal and conditional marginal depend upon disjoint parameter-sets.

$$
D_{KL}(P(x, y) \parallel Q(x, y)) = D_{KL}(P(y) \parallel Q(y)) + E_{P(y)}\left[D_{KL}(P(x | y) \parallel Q(x | y))\right]
$$

2. The relative Kullback-Leibler Divergence (KLD) between two distributions 
$P$ and $Q$ is given by:

$$
D_{KL}(P \parallel Q) = E_P[\log(P)] - E_P[\log(Q)]
$$

Dropping the first term, which is constant with respect to the model 
$Q$, and retaining the negative of the latter term gives us the objective of maximizing the likelihood (or minimizing the negative log-likelihood) of $Q$:

$$
-E_P[\log(Q)]
$$

This is the core for all of maximum likelihood estimation, information criteria, and regression and supervised learning using a negative log-likelihood as its loss function.

### Why EnKF and ES works so well

Let $(x,y)\sim P$ possibly non-Gaussian and assume we have a finite dataset to infer a model $Q$ from.
Arguably, the most important aspects to encode in $Q$, without any other knowledge, are 
the two first moments of $P$.
Having access to the sample covariances of
$\Sigma_{xy}$
and $\Sigma_{y}^{-1}$

> after (maximum likelihood when $n>p$) estimation (minimizing empiricla KLD), the distribution $Q$ minimizing KLD towards $P$ is the Gaussian.

Reviewer: ```suggestion
after (maximum likelihood when $n>p$) estimation (minimizing empirical KLD), the distribution $Q$ minimizing KLD towards $P$ is the Gaussian.
```

Furthermore, the expectation of the bias is an increasing function of the dimensionality of $\theta$, or more precisely the number of degrees of freedom in the statistical estimation / parametrization of $Q$.
The expected bias is decreasing in ensemble/sample size.

To alleviate this we can either adjust for the expectation in the bias, or just pass through a test-set not seen in the fitting of $\theta$.
The latter does not require advanced use of theory (information criteria), and is an "obvious" thing to do.
Thus, using a test dataset, we would like to evaluate different models $Q$, performing slightly different transports.

## Objectives for Kalman-type ensembles

When constraining $Q$ to only encode information on the two first moments, that are learned from data, the KLD
approach yields a Gaussian distribution $Q$ even when $P$ is non-Gaussian.
We call methods using the Kalman-gain in Gaussian transport from prior to posterior as "Kalman-type" ensemble based data assimilation methods.
They implicitly seek to minimize information loss in this transport.

> Different Kalman-type modelling approaches essentially yields different estimates of the Kalman-gain, $K$.

Reviewer: ```suggestion
Different Kalman-type modelling approaches essentially yield different estimates of the Kalman-gain, $K$.
```

This means we may use the same formula, but with a different test-set of samples.
This is a natural objective if explicit estimates of $\Sigma_{x}$, $\Sigma_{xy}$, and $\Sigma_{y}$ are computed in creating
an estimate $K=\Sigma_{xy}\Sigma_{y}^{-1}$.
Importantly, it allows encoding information or structure of how $\Sigma_{x}$, $\Sigma_{xy}$, and $\Sigma_{y}$ should look like,
for instance  $\Sigma_y=H\Sigma_xH^T+\Sigma_{\epsilon}$. Then only $\Sigma_x$ and $H$ needs estimation if they should be unknown.
It could also be that $\Sigma_x$ is diagonal, or is a function of some $\theta$ in a smaller dimensional space.
We would then require fewer degrees of freedom to fit $\Sigma_x(\theta)$.

Note that if covariance estimates are singular (i.e. EnKF when $p>n$) then this objective cannot be used due to the log-determinant and covariance inverse.
Furthermore, we often only have access to the final estimate of the Kalman-gain, not the full covariance structure.

### Gaussian conditional marginal

The joint KLD is a natural objective to evaluate model performance on.

> However, multiple (Kalman) ensemble methods only produce an estiamte $K$ either explicit or explicit,

Reviewer: ```suggestion
However, multiple (Kalman) ensemble methods only produce an estimate $K$ either explicit or implicit,
```

Not sure I understand this sentence.
Do you mean that several different ensemble methods only produce an estimate of $K$ without specifying covariance structure?
What do you mean by explicit / implicit?
